README.md
<!--- ![Build Status](https://travis-ci.org/HsoiEnterprises/HEAnalytics.svg) -->
<!--- ![Cocoapods Compatible](https://img.shields.io/cocoapods/v/HEAnalytics.svg) -->
![License](https://img.shields.io/badge/license-BSD%203--Clause-blue.svg)
![Platform](https://img.shields.io/badge/platform-iOS-lightgrey.svg)

# HEAnalytics
A simple Swift-based framework for iOS app analytics across various and selectable analytics platforms.


# Background and Approach

Apps want analytics, and analytics are inherently app-specific. However, much of the logic for making analytics go is the same across apps, and one bit of complexity to manage is the inevitable desire to track analytics across multiple analytics platforms (Flurry, Google Analytics, Mixpanel, etc.). Having been a part of enough apps/projects that needed such things, and having dealt with some headaches of other approaches, this is something I came up with to try to manage those issues. It's just one way to approach this problem.

## Flexible Platform Choice

Some apps want to use platform X, some way to use Y, some way to use both X and Y. Inevitably those platforms do things different, so the complexity of contending with their API differences should be something the app developer shouldn't have to worry about. Let that all happen within the analytics framework.

As well, try to be as robust as possible in utilizing that platform's API and features, but bridging functionality where required. For example, Flurry has a nice way of tracking event parameters but Google Analytics does not, so `HEAnalytics` does the best it can to provide the event parameters in a useful manner under GAI. On the other side, GAI has a nice way of tracking screen views while Flurry essentially requires you to treat it as another event; `HEAnalytcs` at least tries to make this event easily discernable from the other events.

Need to support a new platform? Subclass `HEAnalyticsPlatform` and add it to your `AnalyticsPlatformConfig.plist`.

Curently supported analytics platforms:

* Flurry
* Google Analytics
* Intercom
* Localytics
* Mixpanel


## Configuration Options

Platform configuration comes by way of a `AnalyticsPlatformConfig.plist` file within the main app bundle. This allows for greater flexibility in build and runtime configuration. For example, a white-label product could create a `AnalyticsPlatformConfig.plist` per product and your build system brings in the proper .plist file at build time. Or perhaps based upon build configuration (Debug vs. Release) you may choose a different configuration (e.g. you have a dev analytics account so your daily dev work doesn't skew your true analytics data).

## Uniformity

Events generally are identified as a string. However this can be a hassle and introduce errors if one isn't careful. The power of Swift enums, along with the suggested design approach of subclassing and centralization, helps minimize the potential for such errors.

## Centralization and Abstraction

Often analytics are approached by directly accessing the API at the point in the app code where the analytic needs to be logged; the API is either directly calling an analytics platform API or calling some basic wrapper. While reasonable, I found over time this approach did not scale nor maintain well. Instead, I prefer to wrap up and abstract away all the gory details of analytics into the analytics wrapper class (`HEAnalytics` or a subclass) so that the client code merely calls one function passing whatever information is relevant to log.

This allows all analytics details to be centralized in one file. This can make tracking and modification easier as the project carries on.

It enables logic for converting the data-to-track into something the analytics APIs can understand to be performed somewhere else so as to keep the client code cleaner and easier to read. As well, many times while a particular analytic event may be unique in its use in the app, many times the parameters to track are similar across events. By encapsulating such logic into a single location, code reuse can be facilitated so you can write the "parameter conversion" code once and multiple event trackers can easily reuse.


# Supported OS and SDK

- Xcode 8+
- Swift 3
- iOS 8

If you need Swift 2.2 support, use v0.6.2.

No effort has been made to make this work on Mac OS X, and there are presently no plans to do so.


# Installation

Installation can be performed as a git submodule.

While I'd like to add Cococapods support, there are issues.

1. `HEAnalytics` is written in Swift, thus Cocoapods forced it into a framework.
2. All of the analytics SDKs are distributed as static libraries (or fake frameworks)
3. The way Cocoapods works, there are problems with dependency resolution and linking due to how Cocoapods works regarding the mixture of frameworks and static libraries.

So, Cocoapods just won't happen right now. 


## Submodule

Add the HEAnalytics source files to your project.  You can omit `HEAnalyticsPlatform` subclasses for platforms you don't intend to support.

You will also need to obtain and integrate the analytics platform SDK of your choice (available separately and many are available via Cocoapods).

<!-- 
## Cocoapods

```ruby
pod "HEAnalytics", :subspecs => ["Flurry", "GoogleAnalytics"]
```
-->

# Usage

## AnalyticsPlatformConfig.plist

Create an `AnalyticsPlatformConfig.plist` file. The structure of this file is fairly simple. The plist root is a dictionary. Each entry in the root dictionary is another dictionary, the key of which is the class name of the `HEAnalyticsPlatform` subclass you want to support – the subclass name is important as `HEAnalytics` uses this to perform a dynamic instantiation of the named class. Within the platform dictionary, place the platform-specific configuration info. Refer to each `HEAnalyticsPlatform[subclass].initializePlatform()` implementation for knowledge of what keys/values are supported.

For example, if you want to support both Flurry and Google Analytics, your file might look like this:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>HEAnalyticsPlatformGAI</key>
	<dict>
		<key>trackingID</key>
		<string>UA-12345678-1</string>
		<key>dispatchInterval</key>
		<integer>15</integer>
	</dict>
	<key>HEAnalyticsPlatformFlurry</key>
	<dict>
		<key>apiKey</key>
		<string>ABC123DEF456GHI789J0</string>
	</dict>
</dict>
</plist>
```

It is up to you and your project needs as to how the plist file is integrated into your project. It may be as simple as adding the file to the project and ensuring it is copied into the app bundle during the "Copy Bundle Resources" build phase. It may be that you add a "Run Script" build phase that copies in a different plist file depending upon the build configuration (i.e. one configuration for debug builds, another for release builds). How the file is added to the main bundle is up to you, just ensure that it is.

**In the Example code, be sure to look at the `AnalyticsPlatformConfig` folder, as well as the `Run Script - HEAnalytics` build phase (and `AppBuildPostflight.rb` build script). This is one way to integrate.**


## Subclassing and Integrating `HEAnalytics` 

While not required, it's **strongly** recommended to subclass `HEAnalytics` and that your application use this subclass.

First, you can make your subclass a singleton. Yes I know the arguments for and against singleton, and I believe this is a case where it's not so evil to use a singleton pattern. However, there is nothing that requires or mandates singleton.

Second, `HEAnalytics` ams to provide a unified abstraction layer for analytics platforms, allowing calling code to be simpler, cleaner, easier to read and maintain. Thus, while clients can certainly use the `HEAnalytics` class directly and `trackData()` directly in client code, it's preferred to put the implementation details into a subclass of `HEAnalytics`. Since analytics are inherently app-specific, this calls for a subclass.

Thus, instead of code like:

```swift
class MyViewController: UIViewController {
    @IBAction func sliderValueDidChange(sender: AnyObject?) {
        if let slider = sender as? UISlider {
            // Do whatever you do with the slider value, like updating your data model.
            myDataObject.value = slider.value
            
            let parameters = ["value": slider.value]
            let data = HEAnalyticsData(category: .Settings, event: "Slider Value Updated", parameters: parameters)
            myHEAnalyticsInstance.trackData(data)
        }
    }
}
```

You should do:

```swift
class MyAppAnalytics: HEAnalytics {
    static let sharedInstance = MyAppAnalytics()

    func trackSliderValue(value: Float) {
        let parameters = ["value": value]
        let data = HEAnalyticsData(category: .Settings, event: "Slider Value Updated", parameters: parameters)
        self.trackData(data)
    }
}
```

Then in your code:

```swift
class MyViewController: UIViewController {
    @IBAction func sliderValueDidChange(sender: AnyObject?) {
        if let slider = sender as? UISlider {
            // Do whatever you do with the slider value, like updating your data model.
            myDataObject.value = slider.value
            
            MyAppAnalytics.sharedInstance.trackSliderValue(slider.value)
        }
    }
    
    func HE_analyticsViewTrackingTitle() -> String {
        return "My Interesting View"
    }
}
```

**See `handleUIApplicationDidFinishLaunchingNotification()` in `HEAnalytics.swift` for a more detailed illustration of the utility of this approach.

This approach:

* Keeps calling code cleaner, easier to read.
* Abstracts away the details.
* Encapsulates all analytics code and logic into a single, centralized location (which becomes a useful reference).
* Improves maintainability.

There's nothing that prevents you from the former approach, but in the author's experience the latter approach is preferrable.


## Startup and Initialization

Because `HEAnalytics` automatically tracks `UIApplicationDelegate` notifications, the best place to start analytics tracking is within `application(application, willFinishLaunchingWithOptions)`. Note! _willFinish_ **NOT** _didFinish_. Starting is simple:

```swift
func application(application: UIApplication, willFinishLaunchingWithOptions launchOptions: [NSObject : AnyObject]?) -> Bool {
    MyAppAnalytics.sharedInstance.start()
    return true
}
```

## Tracking

Event tracking is performed by filling out an `HEAnalyticsData` object and passing it to `HEAnalytics.trackData()`. While one can perform this "raw" at any point in code, again it is recommended to have an app-specific subclass of `HEAnalytics` that performs the heavy lifting (see above).

View tracking can be performed by invoking `HEAnalytics.trackView()`, passing the `UIViewController` you wish to track. Invoking `trackView()` can technically be done anywhere, but makes most sense to be called in your `UIViewController` subclass override of `viewDidAppear()`. To facilitate view tracking, `HEAnalytics` extends `UIViewController` with the `HE_analyticsViewTrackingTitle()` function. This function is intended to provide a stable value for analytics view tracking. By default it returns the `viewController.title` if it is non-nil and non-empty, else returns the name of the `UIViewController` (sub)class. This default behavior is acceptable, but may not always be desired. For example, if your ViewController's title is based upon the contents of the ViewController, that may make it difficult for you to track the view. To counter this, your `UIViewController` subclass can override and implement `HE_analyticsViewTrackingTitle()` and return a known stable title string that is useful for tracking and doesn't interfere with your UI.

```swift
override func viewDidAppear(animated: Bool) {
    super.viewDidAppear(animated)
    MyAppAnalytics.sharedInstance.trackView(self)
}

override func HE_analyticsViewTrackingTitle() -> String {
    return "My Cool ViewController"
}
```


## App Version

If the analytics platform supports setting the app's version, `HEAnalytics` will do so. The version will be synthesized from the `CFBundleShortVersionString` and the `CFBundleVersion` as "`CFBundleShortVersionString`.`CFBundleVersion`". The intention is the `CFBundleShortVersionString` is a public/marketing/semantic version only bumped for releases, and the `CFBundleVersion` is a forever-incrementing build number (integer) which can be bumped during development/testing/releases. Thus you may have a "2.0.1.55" then a "2.0.1.56" and finally "2.0.1.57" is actually what's released to the public, so next version becomes "3.0.0.58".

If you wish to have a different approach, you can subclass the relevant `HEAnalyticsPlatform` and override `appVersion()`.


## Opt-Out

Tracking and privacy are important to users. Many analytics platforms offer a means of opting out, and so `HEAnalytics` provides and API for this.

Some considerations:

- `HEAnalytics` considers "start/stop" and "opt-out" as related but distinct. You can still start if opted out, and depending upon the platform it may or may not result in side-effect behaviors.
- `HEAnalytics` does nothing to manage the "opt-out" state. This is something you should expose to your users somewhere in your GUI (e.g. a `UISwitch`). In doing so, you are responsible for saving and restoring the state, enforcing the state, ensuring `HEAnalytics` is in compliance with the state.


# Contact

## Hsoi Enterprises
- [Website][hsoienterprises-website]
- [Facebook][hsoienterprises-facebook]
- [Twitter (@HsoiEnterprises)][hsoienterprises-twitter]

## Creator
- John C. Daub [(@hsoi)][hsoi-twitter]


# Changes / Release Notes

See included `CHANGELOG.md` file.


# License

BSD 3-clause “New” or “Revised” License. See included "License" file.



[hsoienterprises-website]: http://www.hsoienterprises.com
[hsoienterprises-facebook]: https://www.facebook.com/HsoiEnterprises
[hsoienterprises-twitter]: http://twitter.com/hsoienterprises
[hsoi-twitter]: http://twitter.com/hsoi


CHANGELOG.md
# HEAnalytics CHANGELOG


## v1.0.0 - 2016-09-24

* Conversion to Swift 3 (and Xcode 8). `:props:` to @kreeger for the massive contribution!


## v0.6.1 - 2016-09-24

* Fixed a Localytics tracking bug

Last version to support Swift 2.2


## v0.6 - 2016-03-26

* Xcode 7.3 and Swift 2.2 support.
* Tracking "started" state. Thanx to @kreeger
* Added `HEAnalyticsUser` to track user information. Thanx to @kreeger
* Added Localytics support. Thanx (again) to @kreeger
* Simpilfied `HEAnalyticsPlatform.viewControllerTitle()` because Swift 2.2. We shouldn't NEED that logic anyways (right?)


## v0.5.2 - 2015-09-12

Minor adjustments from the Xcode 7 GM


## v0.5.1 - 2015-09-06

Originally this release was going to bring Cocoapods support, but due to HEAnalytics being written in Swift, Cocoapods requiring Swift-based pods to be in frameworks, that all the analytics SDKs are distributed as static libraries, and that the way Cocoapods works regarding dynamic frameworks and static libraries having issues... well, it won't happen. But changes made were primarily towards that end.

### Added

* This CHANGELOG

### Changed

* Cleaned up access levels.
* Updated example project.
* Improved README.


## v0.5 - 2015-07-16

### Added

* Intercom support


## v0.4 - 2015-05-26

### Added

* Added better custom app versioning support.


## v0.3 - 2015-05-08


### Added

* Sample project

### Changed

* Requires Swift 1.2


## v0.2 - 2015-04-23

### Added

* Mixpanel support


## v0.1 - 2015-04-19

Initial release.


HEAnalytics/HEAnalytics.swift
//
//  HEAnalytics.swift
//
//  Created by hsoi on 4/4/15.
//
//  HEAnalytics - Copyright (c) 2015-2016, Hsoi Enterprises LLC
//  All rights reserved.
//  hsoi@hsoienterprises.com
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions are met:
//
//  * Redistributions of source code must retain the above copyright notice, this
//  list of conditions and the following disclaimer.
//
//  * Redistributions in binary form must reproduce the above copyright notice,
//  this list of conditions and the following disclaimer in the documentation
//  and/or other materials provided with the distribution.
//
//  * Neither the name of HEAnalytics nor the names of its
//  contributors may be used to endorse or promote products derived from
//  this software without specific prior written permission.
//
//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
//  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
//  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
//  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
//  FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
//  DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
//  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
//  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
//  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

import UIKit

/**
HEAnalytics provides a base API for logging analytics.

While not required, it's **strongly** recommended to subclass `HEAnalytics` and that your application use this subclass.

First, you can make your subclass a singleton. Yes I know the arguments for and against singleton, and I believe this is a case where it's not so evil to use a singleton pattern. However, there is nothing that requires or mandates singleton.

Second, `HEAnalytics` ams to provide a unified abstraction layer for analytics platforms, allowing calling code to be simpler, cleaner, easier to read and maintain. Thus, while clients can certainly use the `HEAnalytics` class directly and `track(data:)` directly in client code, it's preferred to put the implementation details into a subclass of `HEAnalytics`. Since analytics are inherently app-specific, this calls for a subclass.

Thus, instead of code like:

```
class MyViewController: UIViewController {
    @IBAction func sliderValueDidChange(sender: Any?) {
        if let slider = sender as? UISlider {
            // Do whatever you do with the slider value, like updating your data model.
            myDataObject.value = slider.value

            let parameters = ["value": slider.value]
            let data = HEAnalyticsData(category: .Settings, event: "Slider Value Updated", parameters: parameters)
            myHEAnalyticsInstance.track(data: data)
        }
    }
}
```

You should do:

```
class MyAppAnalytics: HEAnalytics {
    static let sharedInstance = MyAppAnalytics()

    func trackSliderValue(value: Float) {
        let parameters = ["value": value]
        let data = HEAnalyticsData(category: .Settings, event: "Slider Value Updated", parameters: parameters)
        self.track(data: data)
    }
}
```

Then in your code:

```
class MyViewController: UIViewController {
    @IBAction func sliderValueDidChange(sender: Any?) {
        if let slider = sender as? UISlider {
            // Do whatever you do with the slider value, like updating your data model.
            myDataObject.value = slider.value

            MyAppAnalytics.sharedInstance.trackSliderValue(slider.value)
        }
    }

    override var HE_analyticsViewTrackingTitle: String {
        return "My Interesting View"
    }
}
```

This approach:

* Keeps calling code cleaner, easier to read.
* Abstracts away the details.
* Encapsulates all analytics code and logic into a single, centralized location (which becomes a useful reference).
* Improves maintainability.

There's nothing that prevents you from the former approach, but in the author's experience the latter approach is preferrable.
*/
open class HEAnalytics: NSObject {
    
    /// Tracks internal state of if HEAnalytics has been started.
    fileprivate var started = false

    /**
    Designated initializer.
    
    - returns: an instance of HEAnalytics
    */
    public override init() {
        super.init()
    }
    
    
    /**
    Deinitializer.
    
    Will cause analytics collection to `stop()`.
    */
    deinit {
        stop()
    }
    
    
    /// Private record of the instance's `HEAnalyticsPlatform` objects.
    fileprivate var platforms: [HEAnalyticsPlatform] = []
    
    /**
    Starts the recording of analytics.
    
    Loads the `AnalyticsPlatformConfig.plist`, creates the `HEAnalyticsPlatform`s from it, initializes and starts each platform, and registers for some events that it can automatically track for you.
    
    Recommended to be invoked from `application(_:willFinishLaunchingWithOptions:)`.
    */
    open func start() {
        guard !started else { return }
        
        loadPlatforms()
        
        registerForNotifications()
        
        for platform in platforms {
            platform.start()
        }
        
        started = true
    }
    
    
    /**
    Internal function to load the platforms.
    */
    internal func loadPlatforms() {
        assert(platforms.count == 0, "calling HEAnalytics.start() and there are loaded platforms. How did this happen?")
        
        // Hsoi 2015-04-18 - Load up the configuration.
        //
        // Note the interesting things we must do because Swift isn't the most dynamic of languages. Long live Objective-C!
        // A subtle thing you may miss is that our HEAnalyticsPlatform subclasses all have to declare an `@objc` name for
        // the class in the class declaration -- that's vital for this to work.
        let configFileURL = Bundle.main.url(forResource: "AnalyticsPlatformConfig", withExtension: "plist")
        assert(FileManager.default.fileExists(atPath: configFileURL!.path), "AnalyticsPlatformConfig.plist does not exist in the mainBundle")
        
        if let theURL = configFileURL, let configDict = NSDictionary(contentsOf: theURL) {
            configDict.enumerateKeysAndObjects({ (key:Any, value:Any, stop:UnsafeMutablePointer<ObjCBool>) -> Void in
                if let keyString = key as? String, let valueDictionary = value as? [String:Any] {
                    let theClass = NSClassFromString(keyString) as! HEAnalyticsPlatform.Type
                    let platform = theClass.init(platformData: valueDictionary)
                    self.platforms.append(platform)
                }
            })
        }
        assert(platforms.count > 0, "no analytics platforms were loaded. Is the AnalyticsPlatformConfig.plist present and populated?")
    }
    
    
    /**
    Internal function to unload the platforms.
    */
    internal func unloadPlatforms() {
        platforms = []
    }
    
    
    /**
    Registers for various `NSNotification`s that we can handle ourselves, and that would be generally useful to clients.
    
    For now, registers for all useful `UIApplication` notifications.
    */
    internal func registerForNotifications() {
        // Hsoi 2015-04-18 - Most apps want to track UIApplication events, so let's just track them automatically. One less thing
        // for you to have to worry about!
        NotificationCenter.default.addObserver(self, selector: #selector(handleUIApplicationBackgroundRefreshStatusDidChangeNotification(_:)), name: NSNotification.Name.UIApplicationBackgroundRefreshStatusDidChange, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(handleUIApplicationDidBecomeActiveNotification(_:)), name: NSNotification.Name.UIApplicationDidBecomeActive, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(handleUIApplicationDidEnterBackgroundNotification(_:)), name: NSNotification.Name.UIApplicationDidEnterBackground, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(handleUIApplicationDidFinishLaunchingNotification(_:)), name: NSNotification.Name.UIApplicationDidFinishLaunching, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(handleUIApplicationUserDidTakeScreenshotNotification(_:)), name: NSNotification.Name.UIApplicationUserDidTakeScreenshot, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(handleUIApplicationWillEnterForegroundNotification(_:)), name: NSNotification.Name.UIApplicationWillEnterForeground, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(handleUIApplicationWillResignActiveNotification(_:)), name: NSNotification.Name.UIApplicationWillResignActive, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(handleUIApplicationWillTerminateNotification(_:)), name: NSNotification.Name.UIApplicationWillTerminate, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(handleUIContentSizeCategoryDidChangeNotification(_:)), name: NSNotification.Name.UIContentSizeCategoryDidChange, object: nil)
    }
    
    
    /**
    Unregisters for notifications.
    */
    internal func unregisterForNotifications() {
        NotificationCenter.default.removeObserver(self)
    }

    
    /**
    Stops the recording of analytics.
    */
    open func stop() {
        guard started else { return }
        
        for platform in platforms {
            platform.stop()
        }
        
        unregisterForNotifications()
        unloadPlatforms()
        started = false
    }
    
    
    /**
    Sets the "opt-out" feature.
    
    This doesn't affect start/stop, just an opt-out feature. So things may be running, just not recorded. Can depend upon the particulars of an SDK.
    
    NB: this feature is something you generally should expose in the GUI of the app to allow the user to configure if data can be collected or not -- see the terms of use of each analytics platform. HEAnalytics does nothing to manage this setting/feature: it's upon the app developer to expose, maintain, persist, enforce, etc. this setting.
    
    - parameter opt: true to opt out, false to not. Defaults to not being opted out (false)
    */
    open func optOut(_ opt: Bool) {
        for platform in platforms {
            platform.optOut = opt
        }
    }

    
    /**
    The key function for actually tracking the analytics data.
    
    While in general you could fill out an HEAnalyticsData and call track(data:) from client code, the general recommendation is in your HEAnalytics subclass to implement functions for particular events and encapsulate the logic in the analytics class, not client code.
    
    For example:
    
        func tappedSomeInterestingButton() {
            let data = HEAnalyticsData(category: .General, event: "Tapping Interesting Button")
            track(data: data)
        }
    
    then client code just invokes `MyAnalytics.sharedInstance().tappedSomeInterestingButton()`.
    
    - parameter data: The HEAnalyticsData containing the data to track.
    */
    open func track(data: HEAnalyticsData) {
        for platform in platforms {
            platform.track(data: data)
        }
    }
    
    
    /**
    The key function for view tracking.
    
    You'll want to call this at a time such as in the UIViewController's implementation of viewDidAppear(), or some other appropriate location.
    
    - parameter viewController: The UIViewController to track.
    */
    open func track(viewController: UIViewController) {
        for platform in platforms {
            platform.track(viewController: viewController)
        }
    }
    
    
    /**
     The key function for specific user tracking.
     
     You'll want to call this at a time such as after a user signs in, or some other appropriate location.
     
     - parameter user: The HEAnalyticsUser to track.
     */
    open func track(user: HEAnalyticsUser) {
        for platform in platforms {
            platform.track(user: user)
        }
    }
    
    
    /**
     The key function to halt tracking a specific user.
     
     You'll want to call this at a time such as after a user signs out.
     
     - parameter user: The HEAnalyticsUser to stop tracking; optional.
     */
    open func stopTracking(user: HEAnalyticsUser?) {
        for platform in platforms {
            platform.stopTracking(user: user)
        }
    }
    
    
// MARK: - Application Events
    
    
    /**
    Private handler for UIApplicationBackgroundRefreshStatusDidChangeNotification
    
    - parameter notification: the notification object.
    */
    @objc fileprivate func handleUIApplicationBackgroundRefreshStatusDidChangeNotification(_ notification: Notification) {
        var status = "<unknown>"
        let currentBackgroundRefreshStatus = UIApplication.shared.backgroundRefreshStatus
        switch currentBackgroundRefreshStatus {
        case .restricted:
            status = "restricted"
            
        case .denied:
            status = "denied"
            
        case .available:
            status = "available"
        }

        let data = HEAnalyticsData(category: .application, event: "Background Refresh Status Did Change", parameters: ["status":status])
        track(data: data)
    }
    
    
    /**
    Private handler for UIApplicationDidBecomeActiveNotification
    
    - parameter notification: the notification object.
    */
    @objc fileprivate func handleUIApplicationDidBecomeActiveNotification(_ notification: Notification) {
        let data = HEAnalyticsData(category: .application, event: "Did Become Active")
        track(data: data)
    }

    
    /**
    Private handler for UIApplicationDidEnterBackgroundNotification
    
    - parameter notification: the notification object.
    */
    @objc fileprivate func handleUIApplicationDidEnterBackgroundNotification(_ notification: Notification) {
        let data = HEAnalyticsData(category: .application, event: "Did Enter Background")
        track(data: data)
    }
    
    
    /**
    Private handler for UIApplicationDidFinishLaunchingNotification
    
    - parameter notification: the notification object.
    */
    @objc fileprivate func handleUIApplicationDidFinishLaunchingNotification(_ notification: Notification) {
        var parameters = [String:Any]()
        if let notificationObject = notification.object {
            if notificationObject is UIApplication {
                // don't track the application object
            }
            else {
                parameters["object"] = (notificationObject as? NSObject)?.description ?? NSStringFromClass(type(of: notificationObject) as! AnyClass)
            }
        }
        
        if let userInfo = notification.userInfo {
            if let launchOptions = userInfo[UIApplicationLaunchOptionsKey.url] as? URL {
                parameters["launchOptionsURL"] = launchOptions.absoluteString
            }
            if let sourceApplication:Any = userInfo[UIApplicationLaunchOptionsKey.sourceApplication] {
                parameters["sourceApplication"] = sourceApplication
            }
            if let _ = userInfo[UIApplicationLaunchOptionsKey.remoteNotification] {
                parameters["remoteNotification"] = (notification.object as? NSObject)?.description ?? "<unknown>"
            }
            if let _ = userInfo[UIApplicationLaunchOptionsKey.localNotification] {
                parameters["localNotification"] = (notification.object as? NSObject)?.description ?? "<unknown>"
            }
        }
        
        let dataParameters:[String:Any]? = parameters.count != 0 ? parameters : nil
        let data = HEAnalyticsData(category: .application, event: "Did Finish Launching", parameters: dataParameters)
        track(data: data)
    }
    
    
    /**
    Private handler for UIApplicationUserDidTakeScreenshotNotification
    
    - parameter notification: the notification object.
    */
    @objc fileprivate func handleUIApplicationUserDidTakeScreenshotNotification(_ notification: Notification) {
        let data = HEAnalyticsData(category: .application, event: "Did Take Screenshot")
        track(data: data)
    }
    
    
    /**
    Private handler for UIApplicationWillEnterForegroundNotification
    
    - parameter notification: the notification object.
    */
    @objc fileprivate func handleUIApplicationWillEnterForegroundNotification(_ notification: Notification) {
        let data = HEAnalyticsData(category: .application, event: "Will Enter Foreground")
        track(data: data)
    }
    
    
    /**
    Private handler for UIApplicationWillResignActiveNotification
    
    - parameter notification: the notification object.
    */
    @objc fileprivate func handleUIApplicationWillResignActiveNotification(_ notification: Notification) {
        let data = HEAnalyticsData(category: .application, event: "Will Resign Active")
        track(data: data)
    }
    
    
    /**
    Private handler for UIApplicationWillTerminateNotification
    
    - parameter notification: the notification object.
    */
    @objc fileprivate func handleUIApplicationWillTerminateNotification(_ notification: Notification) {
        let data = HEAnalyticsData(category: .application, event: "Will Terminate")
        track(data: data)
    }
    
    
    /**
    Private handler for UIContentSizeCategoryDidChangeNotification
    
    - parameter notification: the notification object.
    */
    @objc fileprivate func handleUIContentSizeCategoryDidChangeNotification(_ notification: Notification) {
        var parameters:[String:Any] = ["contentSize":"<unknown>"]
        if let newSize: Any = notification.userInfo?[UIContentSizeCategoryNewValueKey] {
            parameters["contentSize"] = newSize
        }
        let data = HEAnalyticsData(category: .application, event: "Content Size Category Did Change", parameters: parameters)
        track(data: data)
    }

}

